### PR TITLE
Assignment6

### PR DESCRIPTION
**Pull request by kellanburns**
Worked by myself. Implemented the generics extra credit option.

### CODE REVIEW

File: Calculate/Calculator.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        result = T.Zero;
+
+        if (string.IsNullOrWhiteSpace(calculation))
+            return false;
+
+        var parts = calculation.Split(' ', StringSplitOptions.RemoveEmptyEntries);
+        if (parts.Length != 3)
+            return false;
+
+        var leftText = parts[0];
```
Comment (steeley21) [inline Calculate/Calculator.cs:31]: This is a nitpick, but I'd recommend not using vars here for better readability in tracking the calculation parts.

**Review by steeley21 — COMMENTED** [review]
Instructions

Create a Console project called "Calculate.". ✔
Define a Program Class
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
  is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
      Uses char for the key corresponding to the operators +, -, *, and /. ✔
      Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
  Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
  If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
  Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
  Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔

Extra Credit

Do one of the following two options (or both if you want extra, extra credit) :)

Refactor the redirect portion of the Program class into 'ProgramBase`
Move ProgramBase into a ConsoleUtilities assembly to be used in other console-based projects
Use generics the mathematical operations methods and consider using generic constraints (requires .NET 7.0)✔

Fundamentals

Place all shared project properties into a Directory.Build.props file.
Place all shared project items into a Directory.Build.targets file. (optional)
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔


Nice work!

**Review by thigiang16 — APPROVED** [review]
Create a Console project called "Calculate.". ✔
Define a Program Class
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔
Extra Credit
Do one of the following two options (or both if you want extra, extra credit) :)

Refactor the redirect portion of the Program class into 'ProgramBase`
Move ProgramBase into a ConsoleUtilities assembly to be used in other console-based projects
Use generics the mathematical operations methods and consider using generic constraints (requires .NET 7.0) ✔
Fundamentals
Place all shared project properties into a Directory.Build.props file.
Place all shared project items into a Directory.Build.targets file. (optional)
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

The program and tests are well-structured. I like how the operations are separated into CalculatorOperations, which makes the code easier to read and maintain. One small improvement would be to reduce the use of var for better readability. You could also add an extra test for unsupported operators, negative numbers, extra spaces, or malformed expressions to strengthen coverage. Overall, good work!

**Comment by Hellothereyoko** [comment]
Create a Console project called "Calculate.". ✔
Define a Program Class
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔

Extra Credit

Do one of the following two options (or both if you want extra, extra credit) :)

Refactor the redirect portion of the Program class into 'ProgramBase`
Move ProgramBase into a ConsoleUtilities assembly to be used in other console-based projects
Use generics the mathematical operations methods and consider using generic constraints (requires .NET 7.0)✔

Fundamentals

Place all shared project properties into a Directory.Build.props file.
Place all shared project items into a Directory.Build.targets file. (optional)
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔


One point of praise I'd like to point out: your testing coverage. What methods do you use in ensuring testing coverage? I always feel like I end up missing something there... It'd help me out to pick your brain on this; cuz your coverage rate is pretty dang good! Keep up the good work!

File: Calculate/Program.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    }
+    
+    public static int Main()
+    {
+        var prog = new Program();
+        var calc = new Calculator<int>();
+
+        int runResult = prog.Run<int>(calc);
+        
+        return runResult;
```
Comment (Joshua-Lester3) [inline Calculate/Program.cs:53]: Kind of a nit: I think return codes are supposed to be for status of finished application. I think code 0 is success and anything else is failure. Not something we grade on though :)

File: Calculate.Tests/CalculatorTests.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        bool difference = calc.TryCalculate("6-7", out _);
+        bool product = calc.TryCalculate("6 *7", out _);
+        bool quotient = calc.TryCalculate("six / 7", out _);
+
+        // Using 'IsFalse' here to avoid compiler warnings
+        Assert.IsFalse(sum);
+        Assert.IsFalse(difference);
+        Assert.IsFalse(product);
+        Assert.IsFalse(quotient);
+    }
```
Comment (Joshua-Lester3) [inline Calculate.Tests/CalculatorTests.cs:58]: I'd separate these out into different tests. Same goes for other similar unit tests. Unit tests should usually only have one reason to fail. This has quite a few

**Review by Joshua-Lester3 — COMMENTED** [review]
## Instructions

- Create a _Console_ project called "Calculate". ❌ Doesn't explicitly specify this.
- Define a Program Class
  - Define two init-only setter properties, `WriteLine` and `ReadLine`, that contain delegates for writing a line of text and reading a line of text respectively ✔
  - Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
  - Set the default behavior for the `WriteLine` and `ReadLine` properties to invoke `System.Console` versions of the methods and add an empty default constructor. ❌ not empty
- Define a Calculator class ✔
  - Define static `Add`, `Subtract`, `Multiply`, and `Divide` methods that have two parameters and return a third parameter. ✔
  - Define a read-only property, `MathematicalOperations`, of type `System.Collections.Generics.IReadOnlyDictionary<TKey,TValue>` that:
    - is initialized to a `System.Collections.Generics.Dictionary<<TKey,TValue>` instance that. ✔
      - Uses `char` for the key corresponding to the operators +, -, \*, and /. ✔
      - Has values that correspond with the Add, Subtract, Multiply, and Divide methods. ✔
  - Implement a `TryCalculate` method following "TryParse" pattern ✔
    - Valid `calculation` expressions include such strings as "3 + 4", "42 - 2", etc. ✔
    - If there is no whitespace around the operator, you can assume the `calculation` is invalid and return false. Similarly if the operands are not integers. ✔
    - Use `string.Split()`, pattern matching, logical and operators to parse the string in their entirety ✔
    - Index into the `MathematicalOperations` method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
- Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
- Be sure to use the `WriteLine`/`ReadLine` properties on `Program` for testing the input and output of your program. ✔

## Extra Credit

Do one of the following two options (or both if you want extra, extra credit) :)

- Refactor the redirect portion of the `Program` class into 'ProgramBase` ❌
  - Move ProgramBase into a ConsoleUtilities assembly to be used in other console-based projects
- Use generics the mathematical operations methods and consider using generic constraints (requires .NET 7.0) ✔

## Fundamentals

- Place all shared project properties into a `Directory.Build.props` file.
- Place all shared project items into a `Directory.Build.targets` file. (optional)
- nullable reference types is enabled ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
- and enabled .NET analyzers for both projects ✔
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version) ✔
- All of the above should be unit tested ❌ Not fully unit tested
- Choose simplicity over complexity ✔

File: Calculate/Program.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        WriteLine = writeLine ?? throw new ArgumentNullException(nameof(writeLine));
+        ReadLine = readLine ?? throw new ArgumentNullException(nameof(readLine));
+    }
+
+    public int Run<T>(Calculator<T> calculator) where T : INumber<T>
+    {
+        ArgumentNullException.ThrowIfNull(calculator);
+
+        while (true)
```
Comment (quattro004) [inline Calculate/Program.cs:26]: Consider that having an infinite loop creates a certain kind of interactive experience like a [REPL](https://en.wikipedia.org/wiki/Read%E2%80%93eval%E2%80%93print_loop) vs if you did not have the loop the app would exit on each invocation and could be used like in an automated tool call fashion.

File: Calculate.Tests/ProgramTests.cs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+using System.Numerics;
+
+
+namespace Calculate.Tests;
+
+[TestClass]
+public sealed class ProgramTests
+{
+    [TestMethod]
+    public void Program_Constructor_DefinesNonNullDelegates()
```
Comment (quattro004) [inline Calculate.Tests/ProgramTests.cs:11]: nit: I'm not sure how useful this test is and could probably be removed. I think the other tests will reveal if the program defines the delegates correctly just by using them.

File: Calculate/Calculator.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        var opText = parts[1];
+        var rightText = parts[2];
+
+        if (opText.Length != 1)
+            return false;
+
+        var op = opText[0];
+
+        if (!T.TryParse(leftText, CultureInfo.InvariantCulture, out var left))
```
Comment (quattro004) [inline Calculate/Calculator.cs:40]: nit: The Essential C# guideline says "AVOID omitting braces, except for the simplest of single-line if statements." All these single line if statements make me think what does simplest mean? I mean these are pretty simple but is the one on line 35 simpler?

File: Calculate/CalculatorOperations.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
+using System;
+using System.Numerics;
+
+namespace Calculate;
+
+public static class CalculatorOperations
```
Comment (quattro004) [inline Calculate/CalculatorOperations.cs:6]: I wonder if we need this extra class or if putting this in the Calculator class itself makes more sense?

**Review by quattro004 — APPROVED** [review]
This is looking good overall just a few minor comments and suggestions. Nice work.
